### ISSUE TITLE
More search engine verification woes.

### ISSUE THREAD

**Issue by Maikuolan** [CIDRAM/CIDRAM#50]
Caught these:

```
ID: 1374
Script Version: CIDRAM v1.4.0
Date/Time: 2018.02.01 08:16:56 +0800
IP Address: 141.8.143.181
Hostname: 141-8-143-181.spider.yandex.com
Signatures Count: 1
Signatures Reference: outgen.php:L240
Why Blocked: Fake YandexBot!
User Agent: Mozilla/5.0 (compatible; YandexBot/3.0; +http://yandex.com/bots)
Reconstructed URI: http://www.[REDACTED].com/painting/still-life
reCAPTCHA State: Enabled.

ID: 1375
Script Version: CIDRAM v1.4.0
Date/Time: 2018.02.01 08:18:54 +0800
IP Address: 141.8.143.181
Hostname: 141-8-143-181.spider.yandex.com
Signatures Count: 1
Signatures Reference: outgen.php:L240
Why Blocked: Fake YandexBot!
User Agent: Mozilla/5.0 (compatible; YandexBot/3.0; +http://yandex.com/bots)
Reconstructed URI: http://www.[REDACTED].com/painting/abstract
reCAPTCHA State: Enabled.

ID: 1376
Script Version: CIDRAM v1.4.0
Date/Time: 2018.02.01 10:05:14 +0800
IP Address: 141.8.143.181
Hostname: 141-8-143-181.spider.yandex.com
Signatures Count: 1
Signatures Reference: outgen.php:L240
Why Blocked: Fake YandexBot!
User Agent: Mozilla/5.0 (compatible; YandexBot/3.0; +http://yandex.com/bots)
Reconstructed URI: http://www.[REDACTED].com/robots.txt
reCAPTCHA State: Enabled.

ID: 1380
Script Version: CIDRAM v1.4.0
Date/Time: 2018.02.01 13:04:47 +0800
IP Address: 141.8.143.181
Hostname: 141-8-143-181.spider.yandex.com
Signatures Count: 1
Signatures Reference: outgen.php:L240
Why Blocked: Fake YandexBot!
User Agent: Mozilla/5.0 (compatible; YandexBot/3.0; +http://yandex.com/bots)
Reconstructed URI: http://www.[REDACTED].com/
reCAPTCHA State: Enabled.
```

..And this:

```
ID: 786
Script Version: CIDRAM v1.4.0
Date/Time: 2018.02.01 00:12:08 +0800
IP Address: 157.55.39.56
Hostname: msnbot-157-55-39-56.search.msn.com
Signatures Count: 1
Signatures Reference: 157.55.38.0/23, xcldbyp_cidram.inc:L32, outgen.php:L219
Why Blocked: Cloud service ("msaz_cidram.inc-IPv4", L683:F5), BingBot MSAZ Bypass, Fake Bingbot!
User Agent: Mozilla/5.0 (iPhone; CPU iPhone OS 7_0 like Mac OS X) AppleWebKit/537.51.1 (KHTML, like Gecko) Version/7.0 Mobile/11A465 Safari/9537.53 (compatible; bingbot/2.0; +http://www.bing.com/bingbot.htm)
Reconstructed URI: http://[REDACTED].com/
reCAPTCHA State: Enabled.
```

..today, via some CIDRAM installations running on two different websites that I'm helping to maintain at the moment.

Said blocked requests cite "Fake Bingbot" and "Fake YandexBot" in their Why Blocked fields. In short, these requests shouldn't have been blocked (false positives). I haven't yet investigated this thoroughly, and I'm not yet sure whether this is a bug in the code, some quirk of the hosting service that these websites use or whatever else, and I'm not yet sure why this has happened, but I'll investigate it shortly. Possible bug-fix inbound soon.

**Comment by Maikuolan** [comment]
Thus far, I haven't been successful in replicating the problem locally. 😕

Will continue testing and post results when I've got a moment.

**Comment by DanielRuf** [comment]
Is it related to DNS or other issues? Is this not reproducible even by mocking yourself with the ip, useragent and so on as the said bots?

**Comment by Maikuolan** [comment]
> Is it related to DNS or other issues?

A problem with search engine verification would generally lend itself to being a DNS issue, but in this case, I'm not entirely sure. DNS seems to be resolving correctly, but.. is being ignored, for whatever reason. My instinct tells me it's probably something small and dumb that I've missed, like a wrong operator buried somewhere in the code or something like that, but being thus far unable to reproduce the problem locally makes me think I'm missing something, or maybe something has changed in the server environment that these websites are hosted on or something like that.

> Is this not reproducible even by mocking yourself with the ip, useragent and so on as the said bots?

Yep. At least, not yet. The falsing seems to be inconsistent though; Most are requests are getting through, but some aren't. I'm going to keep trying to reproduce the problem, but if I don't have any luck soon, I'll take a closer look at the installs to see if something has unexpectedly changed, and at the server environment, I think.

**Comment by Maikuolan** [comment]
Still haven't been able to reproduce the problem locally yet, and I haven't yet seen any repeats of these false positives since first posting this issue here.

I'll leave this issue open for a few more days, in case anyone wants to give any input, ideas, suggestions and so on; or even to report if they've had any similar experiences. Otherwise, if nothing changes (i.e., no repeats of these false positives, no success with reproducing them, no reports of similar incidences elsewhere, etc), I might just put it down to server quirks or something I guess and let it go. I don't particularly like not knowing why something has gone wrong when it goes wrong, but I'm not sure how else I could approach this issue at this time.

**Comment by Maikuolan** [comment]
*Closing.*

**Comment by Maikuolan** [comment]
*Removing "Abandoned" label, adding "Work-In-Progress" label, and reopening.*

I think I might've actually just gotten a bit closer to figuring this out today.

If, for example, I try doing a forward/reverse lookup on my own home IP address, I get results like this (actual hostname and IP [REDACTED] for security/privacy):
```
array(9) {
  ["Status"]=>
  int(0)
  ["TC"]=>
  bool(false)
  ["RD"]=>
  bool(true)
  ["RA"]=>
  bool(true)
  ["AD"]=>
  bool(false)
  ["CD"]=>
  bool(false)
  ["Question"]=>
  array(1) {
    [0]=>
    array(2) {
      ["name"]=>
      string(45) "[REDACTED]"
      ["type"]=>
      int(1)
    }
  }
  ["Answer"]=>
  array(1) {
    [0]=>
    array(4) {
      ["name"]=>
      string(45) "[REDACTED]"
      ["type"]=>
      int(1)
      ["TTL"]=>
      int(299)
      ["data"]=>
      string(13) "[REDACTED]"
    }
  }
  ["Comment"]=>
  string(49) "Response from [REDACTED]."
}
```

That's pretty normal, and should work fine for the purpose of search engine verification in CIDRAM.

However, with BingBot, YandexBot, and some others, sometimes, when doing a forward/reverse lookups on their IPs, we get results like this:
```
array(8) {
  ["Status"]=>
  int(3)
  ["TC"]=>
  bool(false)
  ["RD"]=>
  bool(true)
  ["RA"]=>
  bool(true)
  ["AD"]=>
  bool(false)
  ["CD"]=>
  bool(false)
  ["Question"]=>
  array(1) {
    [0]=>
    array(2) {
      ["name"]=>
      string(37) "msnbot-131-253-27-138.search.msn.com."
      ["type"]=>
      int(1)
    }
  }
  ["Authority"]=>
  array(1) {
    [0]=>
    array(4) {
      ["name"]=>
      string(8) "msn.com."
      ["type"]=>
      int(6)
      ["TTL"]=>
      int(450)
      ["data"]=>
      string(67) "ns1.msft.net. msnhst.microsoft.com. 2018031203 7200 900 2419200 900"
    }
  }
}
```

*Note the distinctly different natures of the result sets: The first returns an answer directly, whereas the second does not return an answer directly, but instead, cites a preferred authority intended for such lookups.*

In the codebase currently, in the "DNS-Resolve" closure, we have this:
```PHP
    return $CIDRAM['Cache']['DNS-Forwards'][$Host]['IPAddr'] = empty(
        $Results['Answer'][0]['data']
    ) ? '' : preg_replace('/[^0-9a-f.:]/i', '', $Results['Answer'][0]['data']);
```

In short: The closure anticipates that the lookup will either fail, or return an answer directly. It doesn't anticipate a response citing an alternate authority to ask.

This *might* be the cause of the problems we're seeing here. In any case though, it's a problem that needs solving, so I'm reopening this issue.

**Comment by caxanga334** [comment]
Found this today.
```
ID: 3319
Script Version: CIDRAM v1.5.0
Date/Time: Fri, 30 Mar 2018 03:23:33 -0300
IP Address: 66.249.64.72
Hostname: crawl-66-249-64-72.googlebot.com
Signatures Count: 1
Signatures Reference: functions.php:L1454
Why Blocked: Fake Googlebot!
User Agent: Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)
Reconstructed URI: https://blog.DOMAIN.com/2017/10/11/new-dust2-available/
reCAPTCHA State: Disabled.
```

**Comment by Maikuolan** [comment]
Cheers for the report.

Didn't realise this could affect Google too.. Interesting.

**Comment by Maikuolan** [comment]
I haven't encountered any more of these false positives on any of my test installations or on any of the websites that I'm managing where CIDRAM is installed, so I'll go ahead and mark this as *fixed*. If anyone encounters any more of these types of false positives though, let me know, and I'll reopen the issue.
